File: packages/guards/src/intent-classification.ts
import type {
  Guard,
  GuardResult,
  GuardContext,
} from 'open-guardrail-core';

type Intent = 'question' | 'command' | 'complaint' | 'request' | 'unknown';

interface IntentClassificationOptions {
  action: 'block' | 'warn';
  allowedIntents: Intent[];
}

const IMPERATIVE_STARTS = [
  /^(?:do|make|create|build|run|execute|delete|remove|stop|start|show|list|get|set|put|add|update|send|open|close|find|search|go|tell|give|write|read|check)\b/i,
];

const COMPLAINT_WORDS = [
  /\b(?:terrible|horrible|awful|worst|broken|useless|frustrat|disappoint|unacceptable|ridiculous|angry|furious|annoy|hate|disgust)\b/i,
  /\b(?:problem|issue|bug|error|fail|crash|wrong|bad|poor|slow)\b/i,
];

const REQUEST_PATTERNS = [
  /\b(?:please|could\s+you|would\s+you|can\s+you|i(?:'d| would) like|kindly)\b/i,
];

function classifyIntent(text: string): Intent {
  const trimmed = text.trim();

  if (/\?\s*$/.test(trimmed)) return 'question';

  for (const p of REQUEST_PATTERNS) {
    if (p.test(trimmed)) return 'request';
  }

  let complaintScore = 0;
  for (const p of COMPLAINT_WORDS) {
    if (p.test(trimmed)) complaintScore++;
  }
  if (complaintScore >= 2) return 'complaint';

  for (const p of IMPERATIVE_STARTS) {
    if (p.test(trimmed)) return 'command';
  }

  return 'unknown';
}

export function intentClassification(options: IntentClassificationOptions): Guard {
  return {
    name: 'intent-classification',
    version: '0.1.0',
    description: 'Classify user intent and optionally restrict',
    category: 'content',
    supportedStages: ['input'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const intent = classifyIntent(text);
      const allowed = options.allowedIntents.includes(intent);

      return {
        guardName: 'intent-classification',
        passed: allowed,
        action: allowed ? 'allow' : options.action,
        message: allowed
          ? undefined
          : `Intent "${intent}" is not allowed`,
        latencyMs: Math.round(performance.now() - start),
        details: { intent, allowedIntents: options.allowedIntents },
      };
    },
  };
}

File: python/open_guardrail/guards/intent_classification.py
"""Classify user intent and restrict."""
from __future__ import annotations

import re
import time
from typing import List

from open_guardrail.core import GuardResult

_IMPERATIVE = re.compile(
    r"^(?:do|make|create|build|run|execute|delete"
    r"|remove|stop|start|show|list|get|set|put"
    r"|add|update|send|open|close|find|search"
    r"|go|tell|give|write|read|check)\b",
    re.I,
)

_COMPLAINT = [
    re.compile(
        r"\b(?:terrible|horrible|awful|worst|broken"
        r"|useless|frustrat|disappoint|unacceptable"
        r"|ridiculous|angry|furious|annoy|hate)\b",
        re.I,
    ),
    re.compile(
        r"\b(?:problem|issue|bug|error|fail"
        r"|crash|wrong|bad|poor|slow)\b",
        re.I,
    ),
]

_REQUEST = re.compile(
    r"\b(?:please|could\s+you|would\s+you"
    r"|can\s+you|i(?:'d| would) like|kindly)\b",
    re.I,
)


def _classify(text: str) -> str:
    t = text.strip()
    if t.endswith("?"):
        return "question"
    if _REQUEST.search(t):
        return "request"
    score = sum(1 for p in _COMPLAINT if p.search(t))
    if score >= 2:
        return "complaint"
    if _IMPERATIVE.match(t):
        return "command"
    return "unknown"


class _IntentClassification:
    def __init__(
        self,
        *,
        action: str = "block",
        allowed_intents: List[str],
    ) -> None:
        self.name = "intent-classification"
        self.action = action
        self._allowed = allowed_intents

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        intent = _classify(text)
        allowed = intent in self._allowed
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="intent-classification",
            passed=allowed,
            action="allow" if allowed else self.action,
            message=(
                None
                if allowed
                else f'Intent "{intent}" not allowed'
            ),
            latency_ms=round(elapsed, 2),
            details={
                "intent": intent,
                "allowed_intents": self._allowed,
            },
        )


def intent_classification(
    *,
    action: str = "block",
    allowed_intents: List[str],
) -> _IntentClassification:
    return _IntentClassification(
        action=action,
        allowed_intents=allowed_intents,
    )
